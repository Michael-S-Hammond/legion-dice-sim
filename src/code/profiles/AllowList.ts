import allowListFile from '../../data/allowList.json';

import * as UC from './UpgradeCard';

export enum AllowListName {
    attack = "attack",
    defense = "defense"
}

type AllowList = {
    armament?: Array<string>,
    command?: Array<string>,
    comms?: Array<string>,
    counterpart?: Array<string>,
    crew?: Array<string>,
    force?: Array<string>,
    gear?: Array<string>,
    generator?: Array<string>,
    grenades?: Array<string>,
    hardpoint?: Array<string>,
    heavyWeapon?: Array<string>,
    ordinance?: Array<string>,
    personnel?: Array<string>,
    pilot?: Array<string>,
    training?: Array<string>,
}

export function isUpgradeInAllowList(upgrade: UC.Upgrade, allowListName: AllowListName) : boolean {
    if(allowListFile[allowListName]) {
        const list = allowListFile[allowListName] as AllowList;
        if(list[upgrade.type]?.includes(upgrade.name)) {
            return true;
        }
    }
    return false;
}
