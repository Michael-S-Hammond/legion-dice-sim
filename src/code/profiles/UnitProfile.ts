import unitsFile from '../../data/units.json';

import * as T from '../Types';

export enum Faction {
    rebel = "rebel",
    empire = "empire",
    separatist = "separatist",
    republic = "republic"
}

export enum Rank {
    commander = "commander",
    operative = "operative",
    corps = "corps",
    specialForces = "specialForces",
    support = "support",
    heavy = "heavy"
}

export enum UnitType {
    creatureTrooper = "creatureTrooper",
    emplacementTrooper = "emplacementTrooper",
    trooper = "trooper",
    wookieTrooper = "wookieTrooper",
    groundVehicle = "groundVehicle",
    repulsorVehicle = "repulsorVehicle"
}

export enum DefenseDie {
    red = "red",
    white = "white"
}

export enum AttackSurge {
    blank = "blank",
    hit = "hit",
    critical = "critical"
}

export enum FixedPosition {
    front = "front",
    rear = "rear"
}

export type WeaponKeywords = {
    critical?: number,
    cumbersome: boolean,
    fixed?: Array<FixedPosition>,
    impact?: number,
    lethal?: number,
    pierce?: number,
    ram?: number,
    suppressive?: boolean
}

export type AttackDice = {
    red: number,
    black: number,
    white: number
}

export type Weapon = {
    name: string,
    minimumRange: number,
    maximumRange?: number,
    dice: AttackDice,
    keywords?: WeaponKeywords
}

export enum UnitUpgrade {
    heavyWeapon = 'heavyWeapon',
    personnel = 'personnel',
    force = 'force',
    command = 'command',
    hardpoint = 'hardpoint',
    gear = 'gear',
    grenades = 'grenades',
    comms = 'comms',
    pilot = 'pilot',
    training = 'training',
    generator = 'generator',
    armament = 'armament',
    crew = 'crew',
    ordinance = 'ordinance'
}

export type RepairUnitKeyword = {
    value: number,
    capacity: number
}

export enum HoverType {
    ground = "ground"
}

export enum Immune {
    blast = "blast",
    melee = "melee",
    pierce = "pierce",
    range1Weapons = "range1Weapons"
}

export enum TransportType {
    closed = "closed",
    open = "open"
}

export type Transport = {
    closed?: number,
    open?: number
}

export type WeakPoint = {
    rear?: number
}

export type UnitKeyword = {
    agile?: number,
    armor?: boolean,
    armorX?: number,
    arsenal?: number,
    charge?: boolean,
    climbingVehicle?: boolean,
    contingencies?: number,
    coordinate?: string,
    cover?: number,
    covertOps?: boolean,
    dangerSense?: number,
    dauntless?: boolean,
    defend?: number,
    deflect?: boolean,
    detachment?: string
    disengage?: boolean,
    duelist?: boolean,
    enrage?: number,
    equip?: Array<string>,
    expertClimber?: boolean,
    fireSupport?: boolean,
    flawed?: boolean,
    fullPivot?: boolean,
    grounded?: boolean,
    guardian?: number,
    gunslinger?: boolean,
    heavyWeaponTeam?: boolean,
    hover?: HoverType,
    immune?: Array<Immune>,
    immunePierce?: boolean,
    impervious?: boolean,
    inconspicuous?: boolean,
    indomitable?: boolean,
    infiltrate?: boolean,
    inspire?: number,
    jump?: number,
    lightTransport?: Transport,
    loadout?: boolean,
    lowProfile?: boolean,
    marksman?: boolean,
    masterOfTheForce?: number,
    nimble?: boolean,
    quickThinking?: boolean,
    ready?: number,
    relentless?: boolean,
    repair?: RepairUnitKeyword,
    reposition?: boolean
    retinue?: string,
    scale?: boolean,
    scout?: number,
    secretMission?: boolean,
    sentinel?: boolean,
    sharpshooter?: number,
    speeder?: number,
    stationary?: boolean,
    tactical?: number,
    takeCover?: number,
    teamwork?: string,
    transport?: Transport,
    uncannyLuck?: number,
    unhindered?: boolean,
    weakPoint?: WeakPoint
}

export type UnitProfile = {
    faction: Faction,
    name: string,
    subtitle?: string,
    unique?: boolean,
    rank: Rank,
    miniCount: number,
    points: number,
    unitType: UnitType,
    defenseDie: DefenseDie,
    wounds: number,
    courage?: number,
    resilience?: number,
    attackSurge: AttackSurge,
    defenseSurge: boolean,
    speed: number,
    weapons: Array<Weapon>,
    upgrades?: Array<UnitUpgrade>,
    keywords?: UnitKeyword
}

const units: Array<UnitProfile> = <Array<UnitProfile>>(unitsFile.units);

export function getUnits() : Array<UnitProfile> {
    return units;
}

export function convertUnitProfileSurgeToAttackSurge(surge: AttackSurge) : T.AttackSurgeConversion {
    switch(surge) {
        case AttackSurge.critical:
            return T.AttackSurgeConversion.Critical;
        case AttackSurge.hit:
            return T.AttackSurgeConversion.Hit;
        default:
            return T.AttackSurgeConversion.Blank;
    }
}
