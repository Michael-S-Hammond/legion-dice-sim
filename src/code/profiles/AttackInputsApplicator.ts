import * as T from '../Types';
import * as UP from './UnitProfile';
import * as UC from './UpgradeCard';

type Tracking = {
    weaponCount: number,
    activeWeapons: number,
    defaultWeapon: UP.Weapon | null,
    useBaseWeapon: boolean,
    minimumRange: number,
    maximumRange?: number,
    fixedPosition: Array<UP.FixedPosition> | null,
    miniCount: number,
    offense: T.OffenseInput
}

function getBoolean(value: boolean | null | undefined) : boolean {
    return value ? true : false;
}

function isPositionCompatible(weapon: UP.Weapon, position: Array<UP.FixedPosition> | null) : boolean {
    if(weapon.keywords?.fixed && position) {
        for(const arc of weapon.keywords.fixed) {
            if(position.includes(arc)) {
                return true;
            }
        }
        return false;
    }
    return true;
}

function getPositionOverlap(weapon: UP.Weapon, position: Array<UP.FixedPosition> | null) : Array<UP.FixedPosition> | null {
    if(weapon.keywords?.fixed && position) {
        const intersection: Array<UP.FixedPosition> = [];
        for(const arc of weapon.keywords.fixed) {
            if(position.includes(arc)) {
                intersection.push(arc);
            }
        }

        return intersection
    } else if(position) {
        return position;
    } else if(weapon.keywords?.fixed) {
        return weapon.keywords.fixed;
    }
    return null;
}

function isRangeCompatible(weapon: UP.Weapon, minimumRange: number, maximumRange?: number) : boolean {
    // both have melee
    if(weapon.minimumRange === 0 && minimumRange === 0) {
        return true;
    }

    // both have infinite range
    if(weapon.maximumRange === undefined && maximumRange === undefined) {
        return true;
    }

    // one has infinite range
    if(maximumRange === undefined) {
        return weapon.maximumRange !== undefined && weapon.maximumRange >= minimumRange;
    }

    if(weapon.maximumRange === undefined) {
        return maximumRange >= weapon.minimumRange;
    }

    // test ranges for overlap
    return (weapon.maximumRange >= minimumRange && weapon.maximumRange <= maximumRange) ||
        (weapon.minimumRange >= minimumRange && weapon.minimumRange <= maximumRange) ||
        (maximumRange >= weapon.minimumRange && maximumRange <= weapon.maximumRange) ||
        (minimumRange >= weapon.minimumRange && minimumRange <= weapon.maximumRange);
}

function applyValueXKeyword(value: number | undefined, keyword: T.AbilityX, unitCount = 1) {
    if(value !== undefined) {
        if(keyword.active) {
            keyword.value += (value * unitCount);
        } else {
            keyword.active = true;
            keyword.value = (value * unitCount);
        }
    }
}

function applyUpgradeKeywords(upgrade: UC.Upgrade, tracking: Tracking) {
    if(upgrade.keywords) {
        if(upgrade.keywords.duelist) {
            tracking.offense.duelist = true;
        }
        if(upgrade.keywords.jediHunter) {
            tracking.offense.jediHunter = true;
        }
        if(upgrade.keywords.makashiMastery) {
            tracking.offense.makashiMastery = true;
        }
        applyValueXKeyword(upgrade.keywords.precise, tracking.offense.preciseX);
        applyValueXKeyword(upgrade.keywords.sharpshooter, tracking.offense.sharpshooterX);
    }
}

function applyWeaponKeywords(keywords: UP.WeaponKeywords, multiplier: number, tracking: Tracking) {
    if(keywords.blast) {
        tracking.offense.blast = true;
    }
    applyValueXKeyword(keywords.critical, tracking.offense.criticalX, multiplier);
    if(!keywords.highVelocity) {
        tracking.offense.highVelocity = false;
    } else if(tracking.activeWeapons === 0) {
        tracking.offense.highVelocity = true;
    }
    applyValueXKeyword(keywords.impact, tracking.offense.impactX, multiplier);
    applyValueXKeyword(keywords.ion, tracking.offense.ionX, 1);
    applyValueXKeyword(keywords.lethal, tracking.offense.lethalX, multiplier);
    applyValueXKeyword(keywords.pierce, tracking.offense.pierceX, multiplier);
    applyValueXKeyword(keywords.ram, tracking.offense.ramX, multiplier);
    if(keywords.surgeCrit) {
        tracking.offense.surge = T.AttackSurgeConversion.Critical;
    }
}

function applyWeapon(weapon: UP.Weapon, type: UP.UnitUpgrade | null, multiplier: number, tracking: Tracking) {
    if(tracking.activeWeapons >= tracking.weaponCount && type !== UP.UnitUpgrade.heavyWeapon) {
        return;
    }

    if(!isPositionCompatible(weapon, tracking.fixedPosition)) {
        return;
    }

    if(!isRangeCompatible(weapon, tracking.minimumRange, tracking.maximumRange)) {
        return;
    }

    // update compatible fixed position
    tracking.fixedPosition = getPositionOverlap(weapon, tracking.fixedPosition);

    // update compatible range
    tracking.minimumRange = Math.max(weapon.minimumRange, tracking.minimumRange);
    if(!tracking.maximumRange) {
        tracking.maximumRange = weapon.maximumRange;
    } else if(weapon.maximumRange) {
        tracking.maximumRange = Math.min(weapon.maximumRange, tracking.maximumRange);
    }

    if(tracking.activeWeapons === 0) {
        tracking.offense.redDice = weapon.dice.red * multiplier;
        tracking.offense.blackDice = weapon.dice.black * multiplier;
        tracking.offense.whiteDice = weapon.dice.white * multiplier;
    } else {
        tracking.offense.redDice += weapon.dice.red * multiplier;
        tracking.offense.blackDice += weapon.dice.black * multiplier;
        tracking.offense.whiteDice += weapon.dice.white * multiplier;
    }

    if(weapon.keywords) {
        applyWeaponKeywords(weapon.keywords, multiplier, tracking);
    }

    if(type !== UP.UnitUpgrade.heavyWeapon) {
        tracking.activeWeapons++;
    }
}

function applyWeaponUpgrade(upgrade: UC.WeaponUpgrade, multiplier: number, tracking: Tracking) {
    if(upgrade.keywords) {
        applyUpgradeKeywords(upgrade, tracking);
    }
    if(upgrade.weapon !== undefined) {
        if(upgrade.weapon) {
            applyWeapon(upgrade.weapon, upgrade.type, multiplier, tracking);
        }
    }
}

function applyUpgrade(tracking: Tracking, upgrade: UC.Upgrade) : void {
    switch(upgrade.type) {
        case UP.UnitUpgrade.armament:
            applyWeaponUpgrade(upgrade as UC.WeaponUpgrade, tracking.miniCount, tracking);
            break;
        case UP.UnitUpgrade.command:
            // No commands currently affect attack
            break;
        case UP.UnitUpgrade.crew:
            applyWeaponUpgrade(upgrade as UC.CrewUpgrade, 1, tracking);
            break;
        case UP.UnitUpgrade.gear:
            applyUpgradeKeywords(upgrade, tracking);
            break;
        case UP.UnitUpgrade.generator:
            if(tracking.fixedPosition && tracking.fixedPosition.length > 0) {
                const generator = upgrade as UC.GeneratorUpgrade;
                applyWeaponKeywords(generator.weaponKeywords, tracking.miniCount, tracking);
            }
            break;
        case UP.UnitUpgrade.grenades:
            applyWeaponUpgrade(upgrade as UC.WeaponUpgrade, tracking.miniCount, tracking);
            break;
        case UP.UnitUpgrade.hardpoint:
            applyWeaponUpgrade(upgrade as UC.WeaponUpgrade, tracking.miniCount, tracking);
            break;
        case UP.UnitUpgrade.heavyWeapon:
            applyWeaponUpgrade(upgrade as UC.WeaponUpgrade, 1, tracking);
            break;
        case UP.UnitUpgrade.ordinance:
            applyWeaponUpgrade(upgrade as UC.WeaponUpgrade, tracking.miniCount, tracking);
            break;
        case UP.UnitUpgrade.pilot:
            applyUpgradeKeywords(upgrade, tracking);
            break;
        case UP.UnitUpgrade.training:
            applyUpgradeKeywords(upgrade, tracking);
            break;
    }
}

function createTrackingObject(profile: UP.UnitProfile, weapon: UP.Weapon | null, upgrades: Array<UC.Upgrade>, tokens: T.OffenseTokens) : Tracking {
    let totalWeapons = weapon ? 1 : 0;
    let arsenal = 0;
    if(profile.keywords?.arsenal) {
        arsenal = profile.keywords.arsenal;
    }
    upgrades.forEach(u => {
        if(u.keywords?.arsenal) {
            arsenal += u.keywords.arsenal;
        }

        if(u.type === UP.UnitUpgrade.armament ||
                u.type === UP.UnitUpgrade.grenades ||
                u.type === UP.UnitUpgrade.hardpoint ||
                u.type === UP.UnitUpgrade.ordinance) {
            const weaponUpgrade = <UC.WeaponUpgrade>u;
            if(weaponUpgrade.weapon) {
                totalWeapons++;
            }
        }
    });

    if(arsenal === 0) {
        arsenal = 1;
    }

    return {
        weaponCount: arsenal,
        activeWeapons: 0,
        defaultWeapon: null,
        useBaseWeapon: totalWeapons <= arsenal,
        minimumRange: 0,
        fixedPosition: null,
        miniCount: profile.miniCount,
        offense: {
            redDice: 0,
            blackDice: 0,
            whiteDice: 0,
            surge: UP.convertUnitProfileSurgeToAttackSurge(profile.attackSurge),
            tokens: tokens,
            blast: false,
            criticalX: { active: false, value: 1},
            duelist: getBoolean(profile.keywords?.duelist),
            highVelocity: false,
            impactX: { active: false, value: 1},
            ionX: { active: false, value: 1},
            jediHunter: getBoolean(profile.keywords?.jediHunter),
            lethalX: { active: false, value: 1},
            makashiMastery: getBoolean(profile.keywords?.makashiMastery),
            pierceX: { active: false, value: 1},
            preciseX: profile.keywords?.precise ? { active: true, value: profile.keywords.precise } : { active: false, value: 1},
            ramX: { active: false, value: 1},
            sharpshooterX: { active: false, value: 1}
        }
    }
}

export function createAttackInputsFromProfile(profile: UP.UnitProfile, weapon: UP.Weapon | null, upgrades: Array<UC.Upgrade>, tokens: T.OffenseTokens) : T.AttackInput {
    const tracking = createTrackingObject(profile, weapon, upgrades, tokens);

    if(tracking.useBaseWeapon && weapon) {
        applyWeapon(weapon, null, tracking.miniCount, tracking);
        tracking.defaultWeapon = weapon;
    }

    upgrades.forEach(u => {
        if(u.type !== UP.UnitUpgrade.generator) {
            applyUpgrade(tracking, u);
        }
    });

    upgrades.forEach(u => {
        if(u.type === UP.UnitUpgrade.generator) {
            applyUpgrade(tracking, u);
        }
    })

    const input = T.createDefaultAttackInput();
    input.offense = tracking.offense;
    input.combat.meleeAttack = tracking?.maximumRange !== undefined ? tracking?.maximumRange === 0 : false;

    return input;
}
