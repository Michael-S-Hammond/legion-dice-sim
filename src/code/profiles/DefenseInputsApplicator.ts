import * as T from '../Types';
import * as UP from './UnitProfile';
import * as UC from './UpgradeCard';

type Tracking = {
    defense: T.DefenseInput
}

function getBoolean(value: boolean | null | undefined) : boolean {
    return value ? true : false;
}

function getValueX(value: number | null | undefined) : T.AbilityX {
    if(value !== null && value !== undefined) {
        if(value > 0) {
            return { active: true, value: value }
        }
    }
    return { active: false, value: 1};
}

function applyValueXKeyword(value: number | undefined, keyword: T.AbilityX, unitCount = 1) {
    if(value !== undefined) {
        if(keyword.active) {
            keyword.value += (value * unitCount);
        } else {
            keyword.active = true;
            keyword.value = (value * unitCount);
        }
    }
}

function applyUpgrade(upgrade: UC.Upgrade, tracking: Tracking) {
    if(upgrade.keywords) {
        if(upgrade.keywords.armor) {
            tracking.defense.armor = true;
        }
        applyValueXKeyword(upgrade.keywords.armorX, tracking.defense.armorX);
        if(upgrade.keywords.block) {
            tracking.defense.block = true;
        }

        if(upgrade.keywords.cover === T.Cover.Heavy) {
            tracking.defense.cover = T.Cover.Heavy;
        } else if(upgrade.keywords.cover === T.Cover.Light) {
            tracking.defense.cover = (tracking.defense.cover === T.Cover.None) ? T.Cover.Light : T.Cover.Heavy;
        }

        applyValueXKeyword(upgrade.keywords.dangerSense, tracking.defense.dangerSenseX);
        if(upgrade.keywords.deflect) {
            tracking.defense.deflect = true;
        }
        if(upgrade.keywords.djemSoMastery) {
            tracking.defense.djemSoMastery = true;
        }
        if(upgrade.keywords.duelist) {
            tracking.defense.duelist = true;
        }

        if(upgrade.keywords.immune) {
            upgrade.keywords.immune.forEach(i => {
                switch(i) {
                    case UP.UnitImmune.blast:
                        tracking.defense.immuneBlast = true;
                        break;
                    case UP.UnitImmune.meleePierce:
                        tracking.defense.immuneMeleePierce = true;
                        break;
                    case UP.UnitImmune.pierce:
                        tracking.defense.immunePierce = true;
                        break;
                }
            });
        }

        if(upgrade.keywords.impervious) {
            tracking.defense.impervious = true;
        }
        if(upgrade.keywords.lowProfile) {
            tracking.defense.lowProfile = true;
        }
        if(upgrade.keywords.outmaneuver) {
            tracking.defense.outmaneuver = true;
        }
        if(upgrade.keywords.shielded) {
            tracking.defense.tokens.shield += upgrade.keywords.shielded;
        }
        if(upgrade.keywords.soresuMastery) {
            tracking.defense.soresuMastery = true;
        }
        applyValueXKeyword(upgrade.keywords.uncannyLuck, tracking.defense.uncannyLuckX);
    }
}

function applySpecialUpgrades(upgrades: Array<UC.Upgrade>, tracking: Tracking) {
    if(upgrades.filter(u => u.type === UP.UnitUpgrade.training && u.name === "Duck and Cover").length > 0) {
        if(tracking.defense.cover === T.Cover.None) {
            tracking.defense.cover = T.Cover.Light;
        }
    }
}

function createTrackingObject(profile: UP.UnitProfile, upgrades: Array<UC.Upgrade>, tokens: T.DefenseTokens) : Tracking {
    return {
        defense: {
            armor: getBoolean(profile.keywords?.armor),
            armorX: getValueX(profile.keywords?.armorX),
            block: getBoolean(profile.keywords?.block),
            cover: (profile.keywords?.cover) ? profile.keywords.cover : T.Cover.None,
            dangerSenseX: getValueX(profile.keywords?.dangerSense),
            deflect: getBoolean(profile.keywords?.deflect),
            dieColor: profile.defenseDie,
            djemSoMastery: getBoolean(profile.keywords?.djemSoMastery),
            duelist: getBoolean(profile.keywords?.duelist),
            hasForceUpgrades: (profile.upgrades || []).filter(u => u === UP.UnitUpgrade.force).length > 0,
            immuneBlast: (profile.keywords?.immune || []).filter(i => i === UP.UnitImmune.blast).length > 0,
            immuneMeleePierce: (profile.keywords?.immune || []).filter(i => i === UP.UnitImmune.meleePierce).length > 0,
            immunePierce: (profile.keywords?.immune || []).filter(i => i === UP.UnitImmune.pierce).length > 0,
            impervious: getBoolean(profile.keywords?.impervious),
            lowProfile: getBoolean(profile.keywords?.lowProfile),
            outmaneuver: getBoolean(profile.keywords?.outmaneuver),
            soresuMastery: getBoolean(profile.keywords?.soresuMastery),
            surge: profile.defenseSurge,
            tokens: {
                dodge: tokens.dodge,
                observation: tokens.observation,
                shield: profile.keywords?.shielded ? tokens.shield + profile.keywords.shielded : tokens.shield,
                suppression: tokens.suppression,
                surge: tokens.surge
            },
            uncannyLuckX: getValueX(profile.keywords?.uncannyLuck)
        }
    }
}

export function createDefenseInputsFromProfile(profile: UP.UnitProfile, upgrades: Array<UC.Upgrade>, tokens: T.DefenseTokens) : T.AttackInput {
    const tracking = createTrackingObject(profile, upgrades, tokens);

    upgrades.forEach(u => {
        applyUpgrade(u, tracking);
    });

    applySpecialUpgrades(upgrades, tracking);

    const input = T.createDefaultAttackInput();
    input.defense = tracking.defense;

    return input;
}
