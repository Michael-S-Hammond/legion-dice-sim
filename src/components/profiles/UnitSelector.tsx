import React from 'react';

import * as UP from '../../code/profiles/UnitProfile';

type UnitSelectorProps = {
    id: string,
    units: Array<UP.UnitProfile>,
    selectedUnit: UP.UnitProfile,
    onUnitChange: (unit: UP.UnitProfile) => void,
}

type UnitSelectorState = {
    matchingNames: number
}

class UnitSelector extends React.Component<UnitSelectorProps, UnitSelectorState> {
    constructor(props : UnitSelectorProps) {
        super(props);

        this.state = {
            matchingNames: this.getMatchingNameCount(props.selectedUnit.name)
        };
    }

    private getMatchingNameCount(name: string): number {
        const count = this.props.units.filter(u => u.name === name).length;
        return count;
    }

    private doesSubtitleMatch(first: string | undefined, second: string): boolean {
        return (first === undefined && second.length === 0) || first === second;
    }

    private onNameChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
        const newName = e.target.value;
        let newUnit = this.props.selectedUnit;

        const possibleUnits = this.props.units.filter(u => u.name === newName);
        if(possibleUnits.length > 0) {
            newUnit = possibleUnits[0];
        }

        if(newUnit !== this.props.selectedUnit) {
            this.props.onUnitChange(newUnit);
        }
    }

    private onSubtitleChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
        const newSubtitle = e.target.value;
        let newUnit = this.props.selectedUnit;

        const possibleUnits = this.props.units.filter(u => u.name === this.props.selectedUnit.name && this.doesSubtitleMatch(u.subtitle, newSubtitle));
        if(possibleUnits.length > 0) {
            newUnit = possibleUnits[0];
        }

        if(newUnit !== this.props.selectedUnit) {
            this.props.onUnitChange(newUnit);
        }
    }

    componentDidUpdate(prevProps: UnitSelectorProps) : void {
        if(prevProps.selectedUnit !== this.props.selectedUnit) {
            this.setState({
                matchingNames: this.getMatchingNameCount(this.props.selectedUnit.name)
            });
        }
    }

    render() : JSX.Element {
        return (
            <div>
                <div className="d-flex mx-auto">
                    <select
                        id={this.props.id + "-nameSelect"}
                        value={this.props.selectedUnit.name}
                        className="rounded px-2 mx-auto"
                        onChange={this.onNameChange}>
                        { [...new Set(this.props.units.map(u => u.name))].
                            map(name => <option key={name} value={name}>{name}</option>)}
                    </select>
                </div>
                { this.state.matchingNames === 1 && this.props.selectedUnit.subtitle && this.props.selectedUnit.subtitle.length > 0 &&
                    <div key={this.props.id + "-subtitle-one"} className="text-center mt-2 mx-auto">{ this.props.selectedUnit.subtitle }</div>
                }
                { this.state.matchingNames > 1 &&
                    <div className="d-flex mx-auto mt-2">
                        <select
                            id={this.props.id + "-subtitleSelect"}
                            value={this.props.selectedUnit.subtitle}
                            className="rounded px-2 mx-auto"
                            onChange={this.onSubtitleChange}>
                            { this.props.units.filter(u => u.name === this.props.selectedUnit.name).map(u =>
                                <option key={u.name + "-" + u.subtitle} value={u.subtitle}>{u.subtitle}</option>) }
                        </select>
                    </div>
                }
            </div>
        );
    }
}

export default UnitSelector;
