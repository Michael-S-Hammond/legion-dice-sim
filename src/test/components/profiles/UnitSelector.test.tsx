import React from 'react';

import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import * as UP from '../../../code/profiles/UnitProfile';
import UnitSelector from 'components/profiles/UnitSelector';

describe('UnitSelector', () => {
    it('matches the snapshot', () => {
        const onUnitChange = jest.fn();
        const units = UP.getUnits().filter(u => u.faction === UP.Faction.empire && u.rank === UP.Rank.corps);
        const unit = units.filter(u => u.name === 'DF-90 Mortar Trooper')[0];

        const selector = shallow(<UnitSelector
                id='my-unit-selector'
                selectedUnit={unit}
                units={units}
                onUnitChange={onUnitChange}
            ></UnitSelector>);
        const snapshot = shallowToJson(selector);
        expect(snapshot).toMatchSnapshot();
    });

    it('matches the snapshot with subtitle', () => {
        const onUnitChange = jest.fn();
        const units = UP.getUnits().filter(u => u.faction === UP.Faction.rebel && u.rank === UP.Rank.operative);
        const unit = units.filter(u => u.name === 'Luke Skywalker')[0];

        const selector = shallow(<UnitSelector
                id='my-unit-selector'
                selectedUnit={unit}
                units={units}
                onUnitChange={onUnitChange}
            ></UnitSelector>);
        const snapshot = shallowToJson(selector);
        expect(snapshot).toMatchSnapshot();
    });

    it('matches the snapshot when multiple units have same name', () => {
        const onUnitChange = jest.fn();
        const units = UP.getUnits().filter(u => u.faction === UP.Faction.republic && u.rank === UP.Rank.specialForces);
        const unit = units.filter(u => u.name === 'ARC Troopers')[1];

        const selector = shallow(<UnitSelector
                id='my-unit-selector'
                selectedUnit={unit}
                units={units}
                onUnitChange={onUnitChange}
            ></UnitSelector>);
        const snapshot = shallowToJson(selector);
        expect(snapshot).toMatchSnapshot();
    });

    it('handles unit change', () => {
        const units = UP.getUnits().filter(u => u.faction === UP.Faction.rebel && u.rank === UP.Rank.commander);
        let unit = units.filter(u => u.name === 'Luke Skywalker')[0];
        const targetUnit = units.filter(u => u.name === 'Han Solo')[0];

        const onUnitChange = jest.fn(newUnit => {
            unit = newUnit;
        });

        const wrapper = shallow(<UnitSelector
                id='my-unit-selector'
                selectedUnit={unit}
                units={units}
                onUnitChange={onUnitChange}
            ></UnitSelector>);
        wrapper.find('#my-unit-selector-nameSelect').simulate('change',
            { target: { value: 'Han Solo' } });

        expect(onUnitChange).toHaveBeenCalledTimes(1);
        expect(unit).toEqual(targetUnit);
    });

    it('handles unit change to one with multiple of the same name', () => {
        const units = UP.getUnits().filter(u => u.faction === UP.Faction.empire && u.rank === UP.Rank.specialForces);
        let unit = units.filter(u => u.name === 'Imperial Death Troopers')[0];
        const targetUnit = units.filter(u => u.name === 'Imperial Special Forces')[0];

        const onUnitChange = jest.fn(newUnit => {
            unit = newUnit;
        });

        const wrapper = shallow(<UnitSelector
                id='my-unit-selector'
                selectedUnit={unit}
                units={units}
                onUnitChange={onUnitChange}
            ></UnitSelector>);
        wrapper.find('#my-unit-selector-nameSelect').simulate('change',
            { target: { value: 'Imperial Special Forces' } });

        expect(onUnitChange).toHaveBeenCalledTimes(1);
        expect(unit).toEqual(targetUnit);
    });

    it('handles unit change to via subtitle', () => {
        const units = UP.getUnits().filter(u => u.faction === UP.Faction.republic && u.rank === UP.Rank.specialForces);
        let unit = units.filter(u => u.name === 'ARC Troopers')[0];
        const targetUnit = units.filter(u => u.name === 'ARC Troopers')[1];

        const onUnitChange = jest.fn(newUnit => {
            unit = newUnit;
        });

        const wrapper = shallow(<UnitSelector
                id='my-unit-selector'
                selectedUnit={unit}
                units={units}
                onUnitChange={onUnitChange}
            ></UnitSelector>);
        wrapper.find('#my-unit-selector-subtitleSelect').simulate('change',
            { target: { value: 'Strike Team' } });

        expect(onUnitChange).toHaveBeenCalledTimes(1);
        expect(unit).toEqual(targetUnit);
    });

    it('handles updating units', () => {
        const units = UP.getUnits().filter(u => u.faction === UP.Faction.separatist && u.rank === UP.Rank.specialForces);
        let unit = units[1];
        const newUnits = UP.getUnits().filter(u => u.faction === UP.Faction.separatist && u.rank === UP.Rank.operative);
        const targetUnit = newUnits[0];

        const onUnitChange = jest.fn(newUnit => {
            unit = newUnit;
        });

        const selector = shallow(<UnitSelector
                id='my-unit-selector'
                selectedUnit={unit}
                units={units}
                onUnitChange={onUnitChange}
            ></UnitSelector>);
        expect(selector.state('matchingNames')).toEqual(2);
        selector.setProps({
            units: newUnits,
            selectedUnit: targetUnit
        });
        expect(selector.state('matchingNames')).toEqual(1);
    });
});
