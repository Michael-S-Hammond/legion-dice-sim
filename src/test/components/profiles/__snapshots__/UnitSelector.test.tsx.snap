// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnitSelector matches the snapshot 1`] = `
<div>
  <div
    className="d-flex mx-auto"
  >
    <select
      aria-label="Unit name"
      className="rounded px-2 mx-auto"
      id="my-unit-selector-nameSelect"
      onChange={[Function]}
      value="DF-90 Mortar Trooper"
    >
      <option
        key="DF-90 Mortar Trooper"
        value="DF-90 Mortar Trooper"
      >
        DF-90 Mortar Trooper
      </option>
      <option
        key="Shoretroopers"
        value="Shoretroopers"
      >
        Shoretroopers
      </option>
      <option
        key="Snowtroopers"
        value="Snowtroopers"
      >
        Snowtroopers
      </option>
      <option
        key="Stormtroopers"
        value="Stormtroopers"
      >
        Stormtroopers
      </option>
    </select>
  </div>
</div>
`;

exports[`UnitSelector matches the snapshot when multiple units have same name 1`] = `
<div>
  <div
    className="d-flex mx-auto"
  >
    <select
      aria-label="Unit name"
      className="rounded px-2 mx-auto"
      id="my-unit-selector-nameSelect"
      onChange={[Function]}
      value="ARC Troopers"
    >
      <option
        key="ARC Troopers"
        value="ARC Troopers"
      >
        ARC Troopers
      </option>
      <option
        key="Wookiee Warriors"
        value="Wookiee Warriors"
      >
        Wookiee Warriors
      </option>
    </select>
  </div>
  <div
    className="d-flex mx-auto mt-2"
  >
    <select
      aria-label="Unit subtitle"
      className="rounded px-2 mx-auto"
      id="my-unit-selector-subtitleSelect"
      onChange={[Function]}
      value="Strike Team"
    >
      <option
        key="ARC Troopers-undefined"
      />
      <option
        key="ARC Troopers-Strike Team"
        value="Strike Team"
      >
        Strike Team
      </option>
    </select>
  </div>
</div>
`;

exports[`UnitSelector matches the snapshot with subtitle 1`] = `
<div>
  <div
    className="d-flex mx-auto"
  >
    <select
      aria-label="Unit name"
      className="rounded px-2 mx-auto"
      id="my-unit-selector-nameSelect"
      onChange={[Function]}
      value="Luke Skywalker"
    >
      <option
        key="Chewbacca"
        value="Chewbacca"
      >
        Chewbacca
      </option>
      <option
        key="K-2SO"
        value="K-2SO"
      >
        K-2SO
      </option>
      <option
        key="Luke Skywalker"
        value="Luke Skywalker"
      >
        Luke Skywalker
      </option>
      <option
        key="R2-D2"
        value="R2-D2"
      >
        R2-D2
      </option>
      <option
        key="Sabine Wren"
        value="Sabine Wren"
      >
        Sabine Wren
      </option>
    </select>
  </div>
  <div
    className="text-center mt-2 mx-auto"
    key="my-unit-selector-subtitle-one"
  >
    Jedi Knight
  </div>
</div>
`;
